Sometimes it appears that stale files are used when gem install is executed.

Sometimes when using `gem install` that has native extensions, the native extension code is seemingly not installed from the gem package.

For example this:

```
> bundle exec bake gem:install
  Successfully built RubyGem
  Name: fiber-profiler
  Version: 0.1.3
  File: fiber-profiler-0.1.3.gem
 0.07s     info: Bake::Gem::Helper [oid=0xbf4] [ec=0xc08] [pid=46378] [2025-02-14 16:53:39 +1300]
               | gem install pkg/fiber-profiler-0.1.3.gem
ERROR:  Error installing pkg/fiber-profiler-0.1.3.gem:
	ERROR: Failed to build gem native extension.

    current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
/Users/samuel/.rubies/ruby-3.3.5/bin/ruby extconf.rb
checking for rb_fiber_current()... yes
checking for rb_ext_ractor_safe()... yes
creating extconf.h
creating Makefile

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-46379-pd35di sitelibdir\=./.gem.20250214-46379-pd35di clean

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-46379-pd35di sitelibdir\=./.gem.20250214-46379-pd35di
compiling ./fiber/profiler/profiler.c
compiling ./fiber/profiler/time.c
compiling ./fiber/profiler/fiber.c
compiling ./fiber/profiler/capture.c
./fiber/profiler/capture.c:483:17: error: use of undeclared identifier 'call'
  483 |                 if (parent == call) {
      |                               ^
1 error generated.
make: *** [capture.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3 for inspection.
Results logged to /Users/samuel/.gem/ruby/3.3.5/extensions/arm64-darwin-24/3.3.0-static/fiber-profiler-0.1.3/gem_make.out
 2.32s    error: Bake::Command [ec=0xc08] [pid=46378] [2025-02-14 16:53:41 +1300]
               | Failed to execute ["gem", "install", "pkg/fiber-profiler-0.1.3.gem"]: pid 46379 exit 1!
               |   RuntimeError: Failed to execute ["gem", "install", "pkg/fiber-profiler-0.1.3.gem"]: pid 46379 exit 1!
               |   → /Users/samuel/.gem/ruby/3.3.5/gems/bake-gem-0.10.0/lib/bake/gem/shell.rb:22 in `system'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-gem-0.10.0/lib/bake/gem/helper.rb:141 in `install_gem'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-gem-0.10.0/bake/gem.rb:42 in `install'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/lib/bake/recipe.rb:110 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/lib/bake/context.rb:99 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/lib/bake/command/call.rb:32 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/lib/bake/command/top.rb:61 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/samovar-2.3.0/lib/samovar/command.rb:21 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/lib/bake/command.rb:12 in `call'
               |     /Users/samuel/.gem/ruby/3.3.5/gems/bake-0.21.0/bin/bake:29 in `<top (required)>'
               |     /Users/samuel/.gem/ruby/3.3.5/bin/bake:25 in `load'
               |     /Users/samuel/.gem/ruby/3.3.5/bin/bake:25 in `<top (required)>'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli/exec.rb:58 in `load'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli/exec.rb:58 in `kernel_load'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli/exec.rb:23 in `run'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli.rb:455 in `exec'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/command.rb:28 in `run'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/invocation.rb:127 in `invoke_command'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor.rb:527 in `dispatch'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli.rb:35 in `dispatch'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/base.rb:584 in `start'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/cli.rb:29 in `start'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.16/exe/bundle:28 in `block in <top (required)>'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/3.3.0/bundler/friendly_errors.rb:117 in `with_friendly_errors'
               |     /Users/samuel/.rubies/ruby-3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.16/exe/bundle:20 in `<top (required)>'
               |     /Users/samuel/.rubies/ruby-3.3.5/bin/bundle:25 in `load'
               |     /Users/samuel/.rubies/ruby-3.3.5/bin/bundle:25 in `<main>'
```


Now you might say, fair enough, but let's check what code is actually included in that gem package:

```
samuel@sakura ~/D/s/fiber-profiler (main)> cd pkg
samuel@sakura ~/D/s/f/pkg (main)> gem unpack fiber-profiler-0.1.3.gem
Unpacked gem: '/Users/samuel/Developer/socketry/fiber-profiler/pkg/fiber-profiler-0.1.3'
samuel@sakura ~/D/s/f/pkg (main)> sed -n 483p fiber-profiler-0.1.3/ext/fiber/profiler/capture.c
		if (parent == current) {
```

So some how, it's not actually compiling the code in the package!

Just to check again because this seems bananas:

```
> gem install fiber-profiler-0.1.3.gem 
ERROR:  Error installing fiber-profiler-0.1.3.gem:
	ERROR: Failed to build gem native extension.

    current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
/Users/samuel/.rubies/ruby-3.3.5/bin/ruby extconf.rb
checking for rb_fiber_current()... yes
checking for rb_ext_ractor_safe()... yes
creating extconf.h
creating Makefile

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-47330-gp0i1d sitelibdir\=./.gem.20250214-47330-gp0i1d clean

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-47330-gp0i1d sitelibdir\=./.gem.20250214-47330-gp0i1d
compiling ./fiber/profiler/profiler.c
compiling ./fiber/profiler/time.c
compiling ./fiber/profiler/fiber.c
compiling ./fiber/profiler/capture.c
./fiber/profiler/capture.c:483:17: error: use of undeclared identifier 'call'
  483 |                 if (parent == call) {
      |                               ^
1 error generated.
make: *** [capture.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3 for inspection.
Results logged to /Users/samuel/.gem/ruby/3.3.5/extensions/arm64-darwin-24/3.3.0-static/fiber-profiler-0.1.3/gem_make.out
```

Okay, so WHERE is that file coming from - of course it's a previous install of the package that failed to compile:

```
samuel@sakura ~/D/s/f/pkg (main) [1]> cd /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
samuel@sakura ~/.g/r/3/g/f/ext> ls
Makefile    extconf.h   extconf.rb* fiber/      fiber.o     profiler.o  time.o
samuel@sakura ~/.g/r/3/g/f/ext> sed -n 483p fiber/profiler/capture.c
		if (parent == call) {
```

That's the old code! Okay so for some reason, installing the gem is not removing existing files. Let's try trashing that file:

```
samuel@sakura ~/.g/r/3/g/f/ext> rm fiber/profiler/capture.c 
```

Now let's try installing the gem again:

```
samuel@sakura ~/D/s/f/pkg (main)> gem install fiber-profiler-0.1.3.gem
ERROR:  Error installing fiber-profiler-0.1.3.gem:
	ERROR: Failed to build gem native extension.

    current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
/Users/samuel/.rubies/ruby-3.3.5/bin/ruby extconf.rb
checking for rb_fiber_current()... yes
checking for rb_ext_ractor_safe()... yes
creating extconf.h
creating Makefile

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-48797-yiexot sitelibdir\=./.gem.20250214-48797-yiexot clean

current directory: /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3/ext
make DESTDIR\= sitearchdir\=./.gem.20250214-48797-yiexot sitelibdir\=./.gem.20250214-48797-yiexot
compiling ./fiber/profiler/profiler.c
compiling ./fiber/profiler/time.c
compiling ./fiber/profiler/fiber.c
linking shared-object Fiber_Profiler.bundle
clang: error: no such file or directory: 'capture.o'
make: *** [Fiber_Profiler.bundle] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/samuel/.gem/ruby/3.3.5/gems/fiber-profiler-0.1.3 for inspection.
Results logged to /Users/samuel/.gem/ruby/3.3.5/extensions/arm64-darwin-24/3.3.0-static/fiber-profiler-0.1.3/gem_make.out
```

Still not working!

Let's try to remove the entire gem:

```
samuel@sakura ~/D/s/f/pkg (main) [1]> gem uninstall fiber-profiler
Successfully uninstalled fiber-profiler-0.1.3
samuel@sakura ~/D/s/f/pkg (main)> gem install fiber-profiler-0.1.3.gem
Building native extensions. This could take a while...
Successfully installed fiber-profiler-0.1.3
Parsing documentation for fiber-profiler-0.1.3
Installing ri documentation for fiber-profiler-0.1.3
Done installing documentation for fiber-profiler after 0 seconds
1 gem installed
```

So it appears that `gem install` is not completely idempotent if existing files exist.

A solution would be to avoid doing in-tree builds, or to ensure that the build directory is clean before trying to build, before installing a gem that already exists delete the existing directory etc.

--- a/lib/rubygems/request_set.rb
+++ b/lib/rubygems/request_set.rb
@@ -181,13 +181,10 @@ class Gem::RequestSet
 
     # Install requested gems after they have been downloaded
     sorted_requests.each do |req|
-      if req.installed?
+      if req.installed? && @always_install.none? {|spec| spec == req.spec.spec }
         req.spec.spec.build_extensions
-
-        if @always_install.none? {|spec| spec == req.spec.spec }
-          yield req, nil if block_given?
-          next
-        end
+        yield req, nil if block_given?
+        next
       end
 
       spec =

--- a/test/rubygems/helper.rb
+++ b/test/rubygems/helper.rb
@@ -683,6 +683,14 @@ class Gem::TestCase < Test::Unit::TestCase
     path
   end
 
+  def write_dummy_extconf(gem_name)
+    write_file File.join(@tempdir, "extconf.rb") do |io|
+      io.puts "require 'mkmf'"
+      yield io if block_given?
+      io.puts "create_makefile '#{gem_name}'"
+    end
+  end
+
   ##
   # Load a YAML string, the psych 3 way
 

--- a/test/rubygems/test_gem_commands_install_command.rb
+++ b/test/rubygems/test_gem_commands_install_command.rb
@@ -647,17 +647,10 @@ ERROR:  Possible alternatives: non_existent_with_hint
     @cmd.options[:args] = %w[a]
 
     use_ui @ui do
-      # Don't use Dir.chdir with a block, it warnings a lot because
-      # of a downstream Dir.chdir with a block
-      old = Dir.getwd
-
-      begin
-        Dir.chdir @tempdir
+      Dir.chdir @tempdir do
         assert_raise Gem::MockGemUi::SystemExitException, @ui.error do
           @cmd.execute
         end
-      ensure
-        Dir.chdir old
       end
     end
 
@@ -684,17 +677,10 @@ ERROR:  Possible alternatives: non_existent_with_hint
     @cmd.options[:args] = %w[a]
 
     use_ui @ui do
-      # Don't use Dir.chdir with a block, it warnings a lot because
-      # of a downstream Dir.chdir with a block
-      old = Dir.getwd
-
-      begin
-        Dir.chdir @tempdir
+      Dir.chdir @tempdir do
         assert_raise Gem::MockGemUi::SystemExitException, @ui.error do
           @cmd.execute
         end
-      ensure
-        Dir.chdir old
       end
     end
 
@@ -720,17 +706,10 @@ ERROR:  Possible alternatives: non_existent_with_hint
     @cmd.options[:args] = %w[a]
 
     use_ui @ui do
-      # Don't use Dir.chdir with a block, it warnings a lot because
-      # of a downstream Dir.chdir with a block
-      old = Dir.getwd
-
-      begin
-        Dir.chdir @tempdir
+      Dir.chdir @tempdir do
         assert_raise Gem::MockGemUi::SystemExitException, @ui.error do
           @cmd.execute
         end
-      ensure
-        Dir.chdir old
       end
     end
 

--- a/test/rubygems/test_gem_dependency_installer.rb
+++ b/test/rubygems/test_gem_dependency_installer.rb
@@ -382,13 +382,9 @@ class TestGemDependencyInstaller < Gem::TestCase
     FileUtils.mv f1_gem, @tempdir
     inst = nil
 
-    pwd = Dir.getwd
-    Dir.chdir @tempdir
-    begin
+    Dir.chdir @tempdir do
       inst = Gem::DependencyInstaller.new
       inst.install "f"
-    ensure
-      Dir.chdir pwd
     end
 
     assert_equal %w[f-1], inst.installed_gems.map(&:full_name)

--- a/test/rubygems/test_gem_dependency_installer.rb
+++ b/test/rubygems/test_gem_dependency_installer.rb
@@ -519,6 +519,58 @@ class TestGemDependencyInstaller < Gem::TestCase
     assert_equal %w[a-1], inst.installed_gems.map(&:full_name)
   end
 
+  def test_install_local_with_extensions_already_installed
+    pend "needs investigation" if Gem.java_platform?
+    pend "ruby.h is not provided by ruby repo" if ruby_repo?
+
+    @spec = quick_gem "a" do |s|
+      s.extensions << "extconf.rb"
+      s.files += %w[extconf.rb a.c]
+    end
+
+    write_dummy_extconf "a"
+
+    c_source_path = File.join(@tempdir, "a.c")
+
+    write_file c_source_path do |io|
+      io.write <<-C
+        #include <ruby.h>
+        void Init_a() { }
+      C
+    end
+
+    package_path = Gem::Package.build @spec
+    installer = Gem::Installer.at(package_path)
+
+    # Make sure the gem is installed and backup the correct package
+
+    installer.install
+
+    package_bkp_path = "#{package_path}.bkp"
+    FileUtils.cp package_path, package_bkp_path
+
+    # Break the extension, rebuild it, and try to install it
+
+    write_file c_source_path do |io|
+      io.write "typo"
+    end
+
+    Gem::Package.build @spec
+
+    assert_raise Gem::Ext::BuildError do
+      installer.install
+    end
+
+    # Make sure installing the good package again still works
+
+    FileUtils.cp "#{package_path}.bkp", package_path
+
+    Dir.chdir @tempdir do
+      inst = Gem::DependencyInstaller.new domain: :local
+      inst.install package_path
+    end
+  end
+
   def test_install_minimal_deps
     util_setup_gems
 

--- a/test/rubygems/test_gem_installer.rb
+++ b/test/rubygems/test_gem_installer.rb
@@ -1478,12 +1478,7 @@ end
 
     @spec = setup_base_spec
     @spec.extensions << "extconf.rb"
-    write_file File.join(@tempdir, "extconf.rb") do |io|
-      io.write <<-RUBY
-        require "mkmf"
-        create_makefile("#{@spec.name}")
-      RUBY
-    end
+    write_dummy_extconf @spec.name
 
     @spec.files += %w[extconf.rb]
 
@@ -1503,12 +1498,7 @@ end
     @spec = setup_base_spec
 
     @spec.extensions << "extconf.rb"
-    write_file File.join(@tempdir, "extconf.rb") do |io|
-      io.write <<-RUBY
-        require "mkmf"
-        create_makefile("#{@spec.name}")
-      RUBY
-    end
+    write_dummy_extconf @spec.name
 
     @spec.files += %w[extconf.rb]
 
@@ -1539,12 +1529,7 @@ end
   def test_install_user_extension_dir
     @spec = setup_base_spec
     @spec.extensions << "extconf.rb"
-    write_file File.join(@tempdir, "extconf.rb") do |io|
-      io.write <<-RUBY
-        require "mkmf"
-        create_makefile("#{@spec.name}")
-      RUBY
-    end
+    write_dummy_extconf @spec.name
 
     @spec.files += %w[extconf.rb]
 
@@ -1571,15 +1556,13 @@ end
 
     @spec = setup_base_spec
     @spec.extensions << "extconf.rb"
-    write_file File.join(@tempdir, "extconf.rb") do |io|
+    write_dummy_extconf @spec.name do |io|
       io.write <<-RUBY
-        require "mkmf"
 
         CONFIG['CC'] = '$(TOUCH) $@ ||'
         CONFIG['LDSHARED'] = '$(TOUCH) $@ ||'
         $ruby = '#{Gem.ruby}'
 
-        create_makefile("#{@spec.name}")
       RUBY
     end
 
@@ -1618,12 +1601,7 @@ end
 
     @spec = setup_base_spec
     @spec.extensions << "extconf.rb"
-    write_file File.join(@tempdir, "extconf.rb") do |io|
-      io.write <<-RUBY
-        require "mkmf"
-        create_makefile("#{@spec.name}")
-      RUBY
-    end
+    write_dummy_extconf @spec.name
 
     rb = File.join("lib", "#{@spec.name}.rb")
     @spec.files += [rb]
@@ -1663,15 +1641,13 @@ end
 
       @spec.extensions << "extconf.rb"
 
-      write_file File.join(@tempdir, "extconf.rb") do |io|
+      write_dummy_extconf @spec.name do |io|
         io.write <<-RUBY
-          require "mkmf"
 
           CONFIG['CC'] = '$(TOUCH) $@ ||'
           CONFIG['LDSHARED'] = '$(TOUCH) $@ ||'
           $ruby = '#{Gem.ruby}'
 
-          create_makefile("#{@spec.name}")
         RUBY
       end
 
@@ -1698,13 +1674,13 @@ end
     @spec.require_paths = ["."]
     @spec.extensions << "extconf.rb"
 
-    File.write File.join(@tempdir, "extconf.rb"), <<-RUBY
-      require "mkmf"
-      CONFIG['CC'] = '$(TOUCH) $@ ||'
-      CONFIG['LDSHARED'] = '$(TOUCH) $@ ||'
-      $ruby = '#{Gem.ruby}'
-      create_makefile("#{@spec.name}")
-    RUBY
+    write_dummy_extconf @spec.name do |io|
+      io.write <<~RUBY
+        CONFIG['CC'] = '$(TOUCH) $@ ||'
+        CONFIG['LDSHARED'] = '$(TOUCH) $@ ||'
+        $ruby = '#{Gem.ruby}'
+      RUBY
+    end
 
     # empty depend file for no auto dependencies
     @spec.files += %W[depend #{@spec.name}.c].each do |file|

